app/models/file_depot_s3.rb

```diff
@@ -0,0 +1,54 @@
+class FileDepotS3 < FileDepot
+  include AmazonAuthenticationMixin
+  attr_accessor :s3
+
+  def self.uri_prefix
+    "s3"
+  end
+
+  def object_store?
+    true
+  end
+
+  def self.validate_settings(settings)
+     _log.debug("Validating settings [#{settings}]")
+    new(:uri => settings[:uri]).verify_credentials(nil, settings.slice(:username, :password))
+  end
+
+  private
+
+  def create_bucket(directory_path)
+    if (ssa_bucket = s3.bucket(directory_path)).exists?
+       _log.debug("Found backup destination bucket #{directory_path}")
+    else
+       _log.debug("Backup destination bucket #{directory_path} does not exist, creating...")
+       ssa_bucket.create
+       _log.debug("Created backup destination bucket #{directory_path}")
+    end
+    ssa_bucket
+  end
+
+  def upload(source, destination)
+    create_bucket(destination_path) unless s3.bucket(destination_path).exists?
+    _log.info("Uploading file: #{destination} to File Depot: #{name}...")
+    _log.info("Uploading file: #{destination_file}... Complete")
+  end
+
+  def destination_file
+    destination_path.join(file.destination_file_name).to_s
+  end
+
+  def destination_path
+    base_path.join(file.destination_directory)
+  end
+
+  def base_path
```

Comment:
I don't think base path is a good name for this. In fact I would assume the "base" of a URL would mean the directory or URL host, not the file name.

Reply:
See file_depot_ftp.rb.  Same code.